controller: add storage exports to mz_cluster_replica_frontiers (#20696)
MaterializeInc/materialize: doc/user/content/sql/system-catalog/mz_internal.md

### `mz_cluster_replica_frontiers`

The `mz_cluster_replica_frontiers` table describes the per-replica frontiers of
each source, sink, materialized view, index, and subscription in the system,

> as observed from the coordinator.

Reviewer: This source also contains frontiers of on-off SELECT dataflows. Should we just gloss over those here?

Reviewer: I guess strictly speaking these are short-lived indexes, so maybe fine to not go into too much detail.

`mz_compute_frontiers` reports the frontiers known to the active compute
replica, while this table reports the frontiers the coordinator is aware of.
Note also that `mz_compute_frontiers` is restricted to compute objects (indexes,
materialized views, and subscriptions) while this table contains storage objects
(sources and sinks) as well.

At this time, we do not make any guarantees about the freshness of these numbers.

<!-- RELATION_SPEC mz_internal.mz_cluster_replica_frontiers -->
| Field         | Type             | Meaning                                                                                                                                                                                       |
| ------------- | ------------     | --------                                                                                                                                                                                      |
| `replica_id`  | [`text`]         | The ID of a cluster replica.                                                                                                                                                                  |

> | `object_id `  | [`text`]         | The ID of the source, sink, index, materialized view, or subscription. If an index, materialized view, or subscription, corresponds to [`mz_compute_exports.export_id`](#mz_compute_exports). |

Reviewer: Should we just drop the "corresponds to" part, like we do for `mz_frontiers`? It seems like this is now more confusing than it is helpful. For example, people might wonder if for sources and sinks the `object_id` corresponds to nothing. Or why we point out `mz_compute_exports` here, rather than `mz_indexes`/`mz_materialized_views`/`mz_subscriptions`.

| `credits_per_hour`    | [`numeric`]                  | The number of compute credits consumed per hour. Corresponds to [`mz_cluster_replica_sizes.credits_per_hour`](#mz_cluster_replica_sizes). |

### `mz_frontiers`

The `mz_frontiers` table describes the frontiers of each source, sink, table,
materialized view, index, and subscription in the system, as observed from the
coordinator.

For objects that are installed on replicas (e.g., materialized views and
indexes), the `replica_id` field is always non-`NULL`. If an object is installed
on multiple replicas, it has multiple entries describing the frontier on each
individual replica. For objects that are not installed on replicas (e.g.,
tables), the `replica_id` field is `NULL`.

If you don't care about per-replica frontiers, using the
[`mz_global_frontiers`](#mz_global_frontiers) view instead is more convenient.

[`mz_compute_frontiers`](#mz_compute_frontiers) is similar to this table, but
`mz_compute_frontiers` reports the frontiers known to the active compute

> replica, while this table reports the frontiers the coordinator is aware of.

Reviewer: A replica is not aware of the frontiers of other replicas, only its own. So `mz_compute_frontier` only contains the single frontier of the replica you are targeting. Whereas the coordinator/controller knows the frontiers of all replicas.

Also, a replica knows about updates to its frontiers before the controller does.

> replica, while this table reports the frontiers the coordinator is aware of.

Reviewer 2: Can we clarify the wording a bit here? Maybe 
```
`mz_compute_frontiers` reports the frontiers known to the active compute
replica, while this table has the frontiers of all replicas.
```

Note also that `mz_compute_frontiers` is restricted to compute objects (indexes,
materialized views, and subscriptions) while this table contains storage objects
(sources, sinks, and tables) as well.

At this time, we do not make any guarantees about the freshness of these numbers.

<!-- RELATION_SPEC mz_internal.mz_frontiers -->
| Field         | Type             | Meaning                                                                                       |
| ------------- | ------------     | --------                                                                                      |
| `object_id`   | [`text`]         | The ID of the source, sink, table, index, materialized view, or subscription.                 |
| `replica_id`  | [`text`]         | The ID of a cluster replica, or `NULL` if the object is not installed on a replica.           |

> | `time`        | [`mz_timestamp`] | The next timestamp at which the output may change, or `NULL` if all output has been produced. |

Reviewer: This uses the same naming as `mz_compute_frontiers`: https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_compute_frontiers

Generally, a frontier can consist of multiple times. I think `mz_compute_frontier` was initially written with the intention that it would contain one row per time in the frontier. You would then reconstruct a frontier by `SELECT time FROM mz_compute_frontiers WHERE id = ...`. But every frontier that compute and storage export today consists of totally ordered times, so there is either a single or no time in the frontier, which means you either have one or zero rows.

It seems fine to rename to `frontier`, as you suggested. But if we do that, we should also rename the other relations that expose frontiers for consistency. I think this warrants a separate PR.

> | `time`        | [`mz_timestamp`] | The next timestamp at which the output may change, or `NULL` if all output has been produced. |

Reviewer 3: I don't have a strong preference on `frontier` vs `time`, just throwing the idea out in case someone else does. Your explanation of one row per time makes sense to me though and I think justifies calling it `time`.

### `mz_global_frontiers`

> The `mz_global_frontiers` view describes global the frontiers of each source,

Reviewer: Done

> The `mz_global_frontiers` view describes global the frontiers of each source,

Reviewer: I added another paragraph. Hopefully it's clearer now.

| `dropped_at`          | [`timestamp with time zone`] | The time at which the replica was dropped, or `NULL` if it still exists.                                                                  |
| `credits_per_hour`    | [`numeric`]                  | The number of compute credits consumed per hour. Corresponds to [`mz_cluster_replica_sizes.credits_per_hour`](#mz_cluster_replica_sizes). |

### `mz_frontiers`

The `mz_frontiers` table describes the frontiers of each source, sink, table,
materialized view, index, and subscription in the system, as observed from the
coordinator.

For objects that are installed on replicas (e.g., materialized views and
indexes), the `replica_id` field is always non-`NULL`. If an object is installed
on multiple replicas, it has multiple entries describing the frontier on each
individual replica. For objects that are not installed on replicas (e.g.,
tables), the `replica_id` field is `NULL`.

> If you don't care about per-replica frontiers, using the

Reviewer: I think only power users will be able to understand what a frontier is, but for those using this table to track replica progress seems plausible. I don't really know the level of knowledge our users though, so if you think they don't care, that's probably true.

Which parts would you remove? I think most of this is also valuable for internal users, apart from this reference to `mz_global_frontiers` perhaps.

[`mz_global_frontiers`](#mz_global_frontiers) view instead is more convenient.

> [`mz_compute_frontiers`](#mz_compute_frontiers) is similar to this table, but
> `mz_compute_frontiers` reports the frontiers known to the active compute
> replica, while this table reports the frontiers of all replicas. Note also that

Reviewer 4: This is saying that `mz_compute_frontiers` is similar to the `mz_global_frontiers` table?

```suggestion
[`mz_compute_frontiers`](#mz_compute_frontiers) is similar to the `mz_global_frontiers` table, but
`mz_compute_frontiers` reports the frontiers known to the active compute
replica, while the `mz_global_frontiers` table reports the frontiers of all replicas. Note also that
```

Reviewer: Oh no, "this table" is supposed to refer to the table this piece of the docs belong to, i.e. `mz_frontiers`.

> `mz_compute_frontiers` is restricted to compute objects (indexes, materialized
> views, and subscriptions) while this table contains storage objects (sources,
> sinks, and tables) as well.

Reviewer 4: ```suggestion
`mz_compute_frontiers` is restricted to compute objects (indexes, materialized
views, and subscriptions) while the `mz_global_frontiers` table contains storage objects (sources,
sinks, and tables) as well.
```

At this time, we do not make any guarantees about the freshness of these numbers.

<!-- RELATION_SPEC mz_internal.mz_frontiers -->
| Field         | Type             | Meaning                                                                             |
| ------------- | ------------     | --------                                                                            |
| `object_id`   | [`text`]         | The ID of the source, sink, table, index, materialized view, or subscription.       |
| `replica_id`  | [`text`]         | The ID of a cluster replica, or `NULL` if the object is not installed on a replica. |
| `time`        | [`mz_timestamp`] | The next timestamp at which the output may change.                                  |

### `mz_global_frontiers`

> The `mz_global_frontiers` view describes the global the frontiers of each

Reviewer 3: ```suggestion
The `mz_global_frontiers` view describes the global frontiers of each
```

Reviewer 3: Lol, no worries we've all been there.

source, sink, table, materialized view, index, and subscription in the system,
as observed from the coordinator.

For objects that are installed on replicas (e.g., materialized views and

> indexes), the global frontier is the maximum of the per-replica frontiers

Reviewer 3: ```suggestion
indexes), the global frontier is the maximum of the per-replica frontiers.
```

Objects that are not installed on replicas only have a single, global frontier.

> In contrast to [`mz_frontiers`](#mz_frontiers) this view does not contain
> per-replica frontiers and contains a single entry for each object.

Reviewer 4: ```suggestion
In contrast to [`mz_frontiers`](#mz_frontiers), the `mz_global_frontiers` view does not contain
per-replica frontiers and contains a single entry for each object.
```

| `records` | [`numeric`] | The number of records in all arrangements in the dataflow.                   |
| `batches` | [`numeric`] | The number of batches in all arrangements in the dataflow.                   |

### `mz_frontiers`

The `mz_frontiers` table describes the frontiers of each source, sink, table,
materialized view, index, and subscription in the system, as observed from

> the coordinator.

Reviewer: Same concern about the frontiers of one-off SELECT dataflows.
